draft-ietf-quic-transport.md

```diff
@@ -3399,6 +3401,77 @@ blocks all those streams from making progress.  An implementation is therefore
 advised to bundle as few streams as necessary in outgoing packets without losing
 transmission efficiency to underfilled packets.
 
+## RETIRE_CONNECTION_ID Frame {#frame-retire-cid}
+
+The RETIRE_CONNECTION_ID frame (type=0x21) informs the peer that older
+connection IDs are being discarded by the peer and will no longer be recognized.
+
+The frame is as follows:
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         Sequence (i)                        ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+
+The fields in the RETIRE_CONNECTION_ID frame are as follows:
+
+Sequence:
+: Sequence number of the oldest connection ID issued by the sender which the
+  sender will still recognize as associated with the current connection. The
+  sequence number is formatted as a variable-length integer.
+
+Upon receipt of a RETIRE_CONNECTION_ID frame, an endpoint MUST consider
+connection IDs older than the specified sequence number as unusable and MUST NOT
+consider the associated stateless reset tokens to be valid. If an older
```

Comment:
I don't think older CIDs should be considered unusable. I can imagine a use for this to retire a CID that you used to try to probe a path that didn't work. In that case, you might still be using the older CID for your main path, and you just want to throw away the new one.

Reply:
Perhaps, but then you'd need to individually retire old CIDs one at a time.  I'm envisioning this as more of a sliding window -- an implementation is moving the end of the window.  Is there a reason you wouldn't just roll forward to another CID when you send this frame, in that situation?

Reply:
We should be dealing is so few numbers of simultaneous CIDs that it should be just fine to be explicit. I think a sliding window puts to much of a restriction that you are constantly moving the primary path to these new CIDs. I must prefer being explicit, per CID.

Reply:
I'm inclined to think that if you're dealing in small numbers of CIDs, it's fine to just remember all of them.  If you reach the point where you need to abandon old CIDs because you have too many outstanding, then we're already in a scenario where you're using a lot of them.

Reply:
I'm surprised at throwing the path away.  I'd think there are separate criteria to throw away a path, independent of needing new CIDs.  Still, I can see where a reasonable implementation might get into a state where it's always discarding a single CID at a time and perhaps wanting to discard CIDs non-contiguously.

Perhaps this is a question to discuss on-list, rather than in a PR comment?

Reply:
I agree that we can follow up on the migration logic separate from this discussion. For this PR though, I don't think we should use a window style for CIDs. I think it restricts implementations from doing other things.